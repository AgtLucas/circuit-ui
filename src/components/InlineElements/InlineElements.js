import { Children } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';

import { clearfix } from '../../styles/style-helpers';

const fallbackBaseStyles = ({ children, theme }) => {
  const childrenCount = Children.count(children);

  return css`
    label: inline-elements--fallback;

    > * {
      display: block;
      width: 100%;
    }

    ${theme.mq.medium`
      > * {
        float: left;
        width: ${1 / childrenCount * 95}%;
        width: calc(
           (100% - ${childrenCount - 1} * ${theme.spacings.byte}) /
            ${childrenCount}
        );
      }

      ${clearfix};
    `};
  `;
};

const baseStyles = ({ theme }) => css`
  label: inline-elements;
  display: flex;
  flex-direction: column;
  width: 100%;

  > * {
    &:not(:last-of-type) {
      margin-bottom: ${theme.spacings.byte};
    }
  }

  ${theme.mq.medium`
    align-items: center;
    flex-direction: row;

    > * {
      flex-grow: 1;
      width: auto;

      &:not(:last-of-type) {
        margin-bottom: 0;
        margin-right: ${theme.spacings.byte};
      }
    }
  `};
`;

const fallbackInlineMobileStyles = ({ theme, inlineMobile, children }) => {
  if (!inlineMobile) {
    return null;
  }

  const childrenCount = Children.count(children);

  return css`
    label: inline-elements--inline-mobile-fallback;

    ${theme.mq.untilMedium`
      > * {
        float: left;
        width: ${1 / childrenCount * 95}%;
        width: calc(
           (100% - ${childrenCount - 1} * ${theme.spacings.byte}) /
            ${childrenCount}
        );
      }

      ${clearfix};
    `};
  `;
};

const inlineMobileStyles = ({ theme, inlineMobile }) =>
  inlineMobile &&
  css`
    label: inline-elements--inline-mobile;

    ${theme.mq.untilMedium`
      flex-direction: row;
      flex-grow: 1;
      width: auto;

      > * {
        &:not(:last-of-type) {
          margin-bottom: 0;
          margin-right: ${theme.spacings.byte};
        }
      }
    `};
  `;

/**
 * Layout helper that displays elements inline. Useful for
 * form elements.
 */
const InlineElements = styled('div')(
  baseStyles,
  fallbackBaseStyles,
  inlineMobileStyles,
  fallbackInlineMobileStyles
);

InlineElements.propTypes = {
  /**
   * The child elements to be displayed inline.
   */
  children: PropTypes.arrayOf(PropTypes.element).isRequired,
  /**
   * Forces inline display even on mobile.
   */
  inlineMobile: PropTypes.bool
};

InlineElements.defaultProps = {
  inlineMobile: false
};

/**
 * @component
 */
export default InlineElements;
