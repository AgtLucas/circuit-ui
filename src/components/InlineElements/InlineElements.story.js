import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import styled from 'react-emotion';

import withTests from '../../util/withTests';
import InlineElements from './InlineElements';

const Box = styled('div')`
  height: 42px;

  &:nth-of-type(n) {
    background-color: yellow;
  }

  &:nth-of-type(2n) {
    background-color: red;
  }
`;

const Container = styled('div')({
  width: '95vw',
  maxWidth: '600px',
  margin: '0 auto',
  border: '1px solid green',
  padding: '12px'
});

storiesOf('InlineElements', module)
  .addDecorator(withTests('InlineElements'))
  .add(
    'Two InlineElements',
    withInfo()(() => (
      <Container>
        <InlineElements>
          <Box />
          <Box />
        </InlineElements>
      </Container>
    ))
  )
  .add(
    'Three InlineElements',
    withInfo()(() => (
      <Container>
        <InlineElements>
          <Box />
          <Box />
          <Box />
        </InlineElements>
      </Container>
    ))
  )
  .add(
    'Three InlineElements stacked on mobile',
    withInfo()(() => (
      <Container>
        <InlineElements inlineMobile>
          <Box />
          <Box />
          <Box />
        </InlineElements>
      </Container>
    ))
  );
