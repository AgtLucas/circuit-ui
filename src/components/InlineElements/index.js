import InlineElements from './InlineElements';

export default InlineElements;
