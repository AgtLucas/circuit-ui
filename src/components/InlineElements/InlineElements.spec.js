import React from 'react';

import InlineElements from '.';

describe('InlineElements', () => {
  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(<InlineElements />);
    expect(actual).toMatchSnapshot();
  });

  it('should render with inlineMobile styles', () => {
    const actual = create(<InlineElements inlineMobile />);
    expect(actual).toMatchSnapshot();
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(<InlineElements />);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
