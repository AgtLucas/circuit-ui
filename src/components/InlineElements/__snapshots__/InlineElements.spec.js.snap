// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineElements should render with default styles 1`] = `
.circuit-0 > * {
  width: calc( (100% - -1 * 8px) / 0 );
}

.circuit-0 > *:not(:last-of-type) {
  margin-right: 8px;
}

<div
  className="circuit-0 circuit-1"
/>
`;
